integration_tests/offboard_attitude.cpp

```diff
@@ -29,70 +35,161 @@ TEST_F(SitlTest, OffboardAttitudeRate)
+    for (int i=0; i<1000; i++)
+    {
+        if (telemetry->attitude_euler_angle().pitch_deg <= 45.0f)
+        {
+            auto pitch_deg = close_loop(0.0f,45.0f,0.0f,telemetry);
+            offboard->set_attitude_rate({pitch_deg.roll, pitch_deg.pitch, pitch_deg.yaw, pitch_deg.thrust});
+            std::this_thread::sleep_for(std::chrono::milliseconds(3));
+        }
+        else
+        {
+            std::cout<< "Pitch angle achieved" << std::endl;
+            std::cout<< "Roll: "<<telemetry->attitude_euler_angle().roll_deg << " Pitch: " <<telemetry->attitude_euler_angle().pitch_deg << " Yaw: " << telemetry->attitude_euler_angle().yaw_deg << std::endl;
+            break;
+        }
+     }
+
+    // Balancing
+    for (int i=0; i<5000; i++)
+    {
+         if (telemetry->attitude_euler_angle().pitch_deg > 0.2f)
+         {
+             auto pitch_bal = close_loop(0.0f,0.0f,0.0f,telemetry);
+             offboard->set_attitude_rate({pitch_bal.roll, pitch_bal.pitch, pitch_bal.yaw, pitch_bal.thrust});
+             std::this_thread::sleep_for(std::chrono::milliseconds(10));
+         }
+        else 
+         {  
+            std::cout<< "Pitch Balanced" << std::endl;
+            std::cout<< "Roll: "<<telemetry->attitude_euler_angle().roll_deg << " Pitch: " <<telemetry->attitude_euler_angle().pitch_deg << " Yaw: " << telemetry->attitude_euler_angle().yaw_deg << std::endl;
+            break;
+         }
+    }
+    std::cout << "Loop broken" << std::endl;
+
+    //Testing Yaw
+    std::cout << "Yaw at -60 degrees/second" << std::endl;
+    offboard->set_attitude_rate({0.0f, 0.0f, -60.0f, 0.5f});
+    for (int i=0; i<1000; i++)
+    {
+        if (telemetry->attitude_euler_angle().yaw_deg >= -60.0f) // Flipping the comparison operator while turning anti-clock wise. 
+        {
+            auto yaw_deg = close_loop(0.0f,0.0f,-60.0f,telemetry);
+            offboard->set_attitude_rate({yaw_deg.roll, yaw_deg.pitch, yaw_deg.yaw, yaw_deg.thrust});
+            std::this_thread::sleep_for(std::chrono::milliseconds(5));
+        }
+        else
+        {
+            std::cout << "Yaw angle acheived" << std::endl;
+            std::cout<< "Roll: "<<telemetry->attitude_euler_angle().roll_deg << " Pitch: " <<telemetry->attitude_euler_angle().pitch_deg << " Yaw: " << telemetry->attitude_euler_angle().yaw_deg << std::endl;
+            break;
+        }
+     }
+    std::cout<< "Roll: "<<telemetry->attitude_euler_angle().roll_deg << " Pitch: " <<telemetry->attitude_euler_angle().pitch_deg << " Yaw: " << telemetry->attitude_euler_angle().yaw_deg << std::endl;
+
+    // Balancing
+    for (int i=0; i<1000; i++)
+    {
+         if (telemetry->attitude_euler_angle().yaw_deg <= 0.2f)
+         {
+             auto yaw_bal = close_loop(0.0f,0.0f,0.0f,telemetry);
+             offboard->set_attitude_rate({yaw_bal.roll, yaw_bal.pitch, yaw_bal.yaw, yaw_bal.thrust});
+             std::this_thread::sleep_for(std::chrono::milliseconds(10));
```

Comment:
It would be nicer to use the callback for new attitude updates instead of sleeping here but we can do that in a future pull request.

Reply:
Sure, I will work on that later (mostly by next week)